booster/library/Booster/Log.hs

```diff
@@ -100,78 +138,84 @@ logMessage' a =
 logPretty :: (LoggerMIO m, Pretty a) => a -> m ()
 logPretty = logMessage . renderOneLineText . pretty
 
-withContext :: LoggerMIO m => LogContext -> m a -> m a
-withContext c =
+withContext :: LoggerMIO m => SimpleContext -> m a -> m a
+withContext c = withContext_ (CLNullary c)
+
+withContext_ :: LoggerMIO m => CLContext -> m a -> m a
+withContext_ c =
     withLogger
         ( \(Logger l) -> Logger $ l . (\(LogMessage alwaysShown ctxt m) -> LogMessage alwaysShown (c : ctxt) m)
         )
 
-newtype TermCtxt = TermCtxt Int
-
-instance ToLogFormat TermCtxt where
-    toTextualLog (TermCtxt hsh) = "term " <> (showHashHex hsh)
-    toJSONLog (TermCtxt hsh) = object ["term" .= showHashHex hsh]
-
-newtype HookCtxt = HookCtxt Text
-
-instance ToLogFormat HookCtxt where
-    toTextualLog (HookCtxt h) = "hook " <> h
-    toJSONLog (HookCtxt h) = object ["hook" .= h]
-
-instance ToLogFormat Term where
-    toTextualLog t = renderOneLineText $ pretty t
-    toJSONLog t = toJSON $ addHeader $ externaliseTerm t
-
-instance ToLogFormat Text where
-    toTextualLog t = t
-    toJSONLog t = String t
-
 withTermContext :: LoggerMIO m => Term -> m a -> m a
-withTermContext t@(Term attrs _) m = withContext (LogContext $ TermCtxt attrs.hash) $ do
-    withContext "kore-term" $ logMessage t
-    m
+withTermContext t@(Term attrs _) m =
+    withContext_ (CLWithId . CtxTerm . ShortId $ showHashHex attrs.hash) $ do
+        withContext CtxKoreTerm $ logMessage t
+        m
 
 withPatternContext :: LoggerMIO m => Pattern -> m a -> m a
 withPatternContext Pattern{term, constraints} m =
-    let t' = foldl' AndTerm term $ Set.toList $ Set.map coerce constraints
+    let t' = foldl' AndTerm term $ Set.toList $ Set.map coerce constraints -- FIXME
```

Comment:
I noticed this earlier while trying to copy/paste a `KoreJson` pattern from a log into a file for custom requests.
* The output is not correctly sorted (we would need to introduce `#Equals{SortBool, <sort-of-term>}(true, _)` around the predicates)
* That would  mean we cannot use `AndTerm` any more
We should probably go to `KoreJson` first  (externalise the pattern) for the output, but then we need to come up with a stable hash for logging purposes

Reply:
Does it cause issues when internalising? I.e. when you copy/pasted `KoreJson` from a log line and tried to send that to the server it reported an error?

Reply:
The server rejects the predicates having a wrong sort IIRC, but I would have to re-try this to be sure.